tests/scalers/gitlab_runner/gitlab_runner_test.go

```diff
@@ -0,0 +1,503 @@
+//go:build e2e
+// +build e2e
+
+package gitlab_runner_test
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/joho/godotenv"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"k8s.io/client-go/kubernetes"
+
+	. "github.com/kedacore/keda/v2/tests/helper"
+)
+
+// Load environment variables from .env file
+var _ = godotenv.Load("../../.env")
+
+const (
+	testName      = "gitlab-runner-test"
+	gitlabBaseURL = "https://gitlab.com"
+
+	ciFileContent = `stages: [deploy]\ndeploy-job:\n  stage: deploy\n  script: [\"sleep 15\"]`
+)
+
+var (
+	personalAccessToken = os.Getenv("GITLAB_PAT")
+
+	defaultHeaders = map[string]string{
+		"PRIVATE-TOKEN": personalAccessToken,
+		"Content-Type":  "application/json",
+	}
+
+	minReplicaCount = 0
+	maxReplicaCount = 1
+
+	scaledObjectName = fmt.Sprintf("%s-so", testName)
+	testNamespace    = fmt.Sprintf("%s-ns", testName)
+	secretName       = fmt.Sprintf("%s-secret", testName)
+	deploymentName   = fmt.Sprintf("%s-deployment", testName)
+)
+
+type templateData struct {
+	TestNamespace             string
+	SecretName                string
+	DeploymentName            string
+	ScaledObjectName          string
+	ScaledJobName             string
+	MinReplicaCount           string
+	MaxReplicaCount           string
+	Pat                       string
+	GitlabAPIURL              string
+	ProjectID                 string
+	TargetPipelineQueueLength string
+}
+
+const (
+	secretTemplate = `
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{.SecretName}}
+  namespace: {{.TestNamespace}}
+data:
+  personalAccessToken: {{.Pat}}
+`
+	triggerAuthTemplate = `
+apiVersion: keda.sh/v1alpha1
+kind: TriggerAuthentication
+metadata:
+  name: gitlab-trigger-auth
+  namespace: {{.TestNamespace}}
+spec:
+  secretTargetRef:
+    - parameter: personalAccessToken
+      name: {{.SecretName}}
+      key: personalAccessToken
+`
+	deploymentTemplate = `
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: {{.DeploymentName}}
+  namespace: {{.TestNamespace}}
+  labels:
+    app: gitlab-runner
+spec:
+  replicas: 0
+  selector:
+    matchLabels:
+      app: gitlab-runner
+  template:
+    metadata:
+      labels:
+        app: gitlab-runner
+    spec:
+      terminationGracePeriodSeconds: 90
+      containers:
+      - name: gitlab-runner
+        image: busybox
+        command: ["/bin/sh", "-c", "trap 'echo SIGTERM received; exit 0' SIGTERM; while true; do sleep 30; done"]
```

Comment:
Does it make sense creating a real runner using their runner image to check that the job is dequeued? https://hub.docker.com/r/gitlab/gitlab-runner/tags

Reply:
I sat down and tried doing that this evening, but then realized that we actually already kind of had this discussion [here](https://github.com/kedacore/keda/issues/5616#issuecomment-2495663291).

Gitlab already does have a mechanism of autoscaling (creating runner per job)
![image](https://github.com/user-attachments/assets/4f066f6e-524d-4452-993d-b54b75250e39)
don't worry about the token in the image, it's been rotated already.

So I think doesn't make sense to manage that using keda. Again, please correct me if I'm wrong, maybe I simply don't quite understand the usecase here.